test-app.gradle

```diff
@@ -0,0 +1,222 @@
+import org.gradle.initialization.DefaultSettings
+
+import java.nio.file.Paths
+
+def generatedFileName = "TestAppPackageList.java"
+def generatedFilePackage = "com.facebook.react"
+def generatedFileContentsTemplate = """
+package $generatedFilePackage;
+
+import com.facebook.react.ReactPackage;
+import java.util.Arrays;
+import java.util.ArrayList;
+
+{{ packageImports }}
+
+public class TestAppPackageList { 
+  public ArrayList<ReactPackage> getPackages() {
+    return new ArrayList<>(Arrays.<ReactPackage>asList(
+      {{ packageClassInstances }}
+    ));
+  }
+}
+"""
+
+class TestAppPlugin implements Plugin<PluginAware> {
+    private String generatedFileName;
+    private String generatedFilePackage;
+    private String generatedFileContentsTemplate;
+    
+    private String packageName;
+    private ArrayList<HashMap<String, String>> packages = new ArrayList<>();
+    
+    TestAppPlugin(
+        String generatedFileName, 
+        String generatedFilePackage, 
+        String generatedFileContentsTemplate
+    ) {
+        this.generatedFileName = generatedFileName;
+        this.generatedFilePackage = generatedFilePackage;
+        this.generatedFileContentsTemplate = generatedFileContentsTemplate;
+    }
+
+    @Override
+    void apply(PluginAware target) {
+        if (target instanceof Settings) {
+            applyToSettings((Settings) target)
+        } else if (target instanceof Project) {
+            applyToProject((Project) target)
+        }
+    }
+
+    /**
+    * Finds the path of the installed npm package with the given name using Node's
+    * module resolution algorithm, which searches "node_modules" directories up to
+    * the file system root. This handles various cases, including:
+    *
+    *   - Working in the open-source RN repo:
+    *       Gradle: /path/to/react-native/ReactAndroid
+    *       Node module: /path/to/react-native/node_modules/[package]
+    *
+    *   - Installing RN as a dependency of an app and searching for hoisted
+    *     dependencies:
+    *       Gradle: /path/to/app/node_modules/react-native/ReactAndroid
+    *       Node module: /path/to/app/node_modules/[package]
+    *
+    *   - Working in a larger repo (e.g., Facebook) that contains RN:
+    *       Gradle: /path/to/repo/path/to/react-native/ReactAndroid
+    *       Node module: /path/to/repo/node_modules/[package]
+    *
+    * The search begins at the given base directory (a File object). The returned
+    * path is a string.
+    */
+    static def findNodeModulePath(baseDir, packageName) {
```

Comment:
I see this is used in another Gradle file too. Could it be an idea to move it out to a separate util.gradle script and apply it to both the Gradle files?

Reply:
Done!